pinc/ProofProject.inc

```diff
@@ -0,0 +1,354 @@
+{
+    use ProofPage;
+
+    public function __construct(object $project, string $state)
+    {
+        global $pguser;
+
+        if ($state != $project->state) {
+            $err = sprintf(
+                _('Warning: Project "%1$s" is not in state "%2$s"; it is now in state "%3$s".'),
+                $project->nameofwork,
+                project_states_text($state),
+                project_states_text($project->state)
+            );
+            throw new ProjectNotInRequiredStateException($err);
+        }
+
+        $project_round = $project->get_project_available_round();
+        $project_round->validate_user_can_access($pguser);
+
+        $this->project = $project;
+        $this->projectid = $project->projectid;
+        $this->round = $project_round;
+    }
+
+    public function checkout(): array
+    {
+        global $pguser;
+
+        [$this->page_name] = get_available_proof_page_array($this->project, $this->round, $pguser);
+
+        $this->page_state = $this->round->page_out_state;
+        $timestamp = time();
+        $setters = join(", ", [
+            set_col_str("state", $this->page_state),
+            set_col_num($this->round->time_column_name, $timestamp),
+            set_col_str($this->round->user_column_name, $pguser),
+        ]);
+        _Page_UPDATE($this->projectid, $this->page_name, $setters);
+
+        _log_page_event($this->projectid, $this->page_name, 'checkout', $pguser, $this->round);
+        _project_adjust_n_available_pages($this->projectid, -1);
+
+        return $this->render_page_data();
+    }
+}
+
+class ProofProjectPage extends ProofProject
+{
+    use ProofPage;
+
+    public function __construct(object $project, string $state, string $page_name, string $page_state)
+    {
+        global $pguser;
+
+        parent::__construct($project, $state);
+
+        $user_column = $this->round->user_column_name;
+        $sql = sprintf(
+            "
+            SELECT state, $user_column
+            FROM $this->projectid
+            WHERE image = '%s'
+            ",
+            DPDatabase::escape($page_name)
+        );
+        $res = DPDatabase::query($sql);
+        $row = mysqli_fetch_assoc($res);
+
+        $current_page_state = $row["state"];
+        if ($page_state != $current_page_state) {
+            $err = sprintf(
+                _('Warning: Page  "%1$s" is no longer in state "%2$s"; it is now in state "%3$s".'),
+                $page_name,
+                $page_state,
+                $current_page_state
+            );
+            throw new ProjectPageInconsistentStateException($err);
+        }
+
+        if ($row[$user_column] != $pguser) {
+            throw new PageNotOwnedException(_("You do not own this page."));
+        }
+
+        $this->page_name = $page_name;
+        $this->page_state = $current_page_state;
+    }
+
+    public function return_to_round(): void
+    {
+        global $pguser;
+
+        $this->validate_page(false, "returnToRound");
```

Comment:
This should not go in a validator because it depends on the page action, but we could split out the relevant piece of the DPage `_Page_require` function to do this.